internal/telemetry/appinsights.go

```diff
@@ -0,0 +1,67 @@
+package telemetry
+
+import (
+	"fmt"
+	"time"
+
+	"github.com/microsoft/ApplicationInsights-Go/appinsights"
+)
+
+var telemetryClient appinsights.TelemetryClient
+
+func SetTelemetryClient(client appinsights.TelemetryClient) {
+	telemetryClient = client
+}
+
+func SetTelemetryClientFromInstrumentationKey(instrumentationKey string) {
+	telemetryClient = appinsights.NewTelemetryClient(instrumentationKey)
+}
+
+type Telemetry struct {
+	Client appinsights.TelemetryClient
+}
+
+func InitializeAppInsights() {
+	instrumentationKey := "3fdeec77-2951-456f-bcb2-9f003b512a3f"
+	config := appinsights.NewTelemetryConfiguration(instrumentationKey)
+	telemetryClient = appinsights.NewTelemetryClientFromConfig(config)
+	SetTelemetryClient(telemetryClient)
+	// Add a diagnostics listener for printing telemetry messages
+	appinsights.NewDiagnosticsMessageListener(func(msg string) error {
+		fmt.Printf("[%s] %s\n", time.Now().Format(time.UnixDate), msg)
+		return nil
+	})
+}
+
+// TrackCommand tracks a command execution event
+func TrackCommand(command string) {
+	event := appinsights.NewEventTelemetry("command")
+	event.Properties["command"] = command
+	telemetryClient.Track(event)
+}
+
+// TrackSubCommand tracks a sub-command execution event
+func TrackSubCommand(command, subCommand string) {
+	event := appinsights.NewEventTelemetry("sub-command")
+	event.Properties["command"] = command
+	event.Properties["sub-command"] = subCommand
+	telemetryClient.Track(event)
+}
+
+func TrackEvent(eventName string) {
```

Comment:
>TrackEvent [](http://example.com/codeflow?start=5&length=10)

this is going to need either a `map[string]string` or `[]string...` argument so callers can pass custom property values to the event.

map is probably more idiomatic, but some people like string arrays where you pass name and value separately like

`TrackEvent("myevent", "prop1Name", "prop1Value", "prop2Name", "prop2Value')`

Reply:
map[string]string is ideal to have . Changes done accordingly.